### ISSUE TITLE
Details about pretraining on ObMan

### ISSUE THREAD

**Issue by Moddel** [bardiadoosti/HOPE#25]
Hi Bardia Doosti! Thank you for your excellent work. I am trying to reimplement your work and have some questions about details of pretraining on Obman dataset. The whole network can be splitted into two main parts: ResNet10+graph convolution(we name it as part1 ) and the Adaptive graph U-Net.  I think there are 3 options to pretrain on Obman:

1. Pretrain the part1 and Graph U-Net seperately. For part1, use the image from Obman as input and get 2D prediction from part1. The loss is computed between part1’s output and GT 2D points from Obman. For part2, use the GT 2D points as input, the loss is computed between Graph U-Net’s output and GT 3D points.

2. Pretrain the whole network end2end. That is, use image as input and the loss is computed between final 3D points output and GT 3D points.

3. Pretrain only the Graph U-Net.

I don’t know the exact pretraining method used in the paper, could you please clarify this? Also, could you offer other detail parameters for pretraining such as learning rate, optimizer, and epochs? Or just detailed code for pretraining can also help very much.

Thank you again for your work.

**Comment by bardiadoosti** [comment]
Hi Moddel,

We first pre-trained the whole model with ObMan and then we trained the whole model with FPHA dataset. Note that at the beginning of the training of each dataset we pre-trained the Adaptive Graph U-Net with some random noise in input 2D coordinates and then trained the whole network with 2D and 3D GT with loss mentioned in the paper.

To sum up here is the training part:
2D GT --> Adaptive Graph U-Net --> 3D (10000 iterations, lr=0.001, times 0.1 after each 4000 steps)
Image --> HOPE-Net --> init 2D, 2D, 3D (5000 iterations, lr=0.001, times 0.1 after each 1000 steps)
2D GT --> Adaptive Graph U-Net --> 3D (10000 iterations, lr=0.001, times 0.1 after each 4000 steps)
Image --> HOPE-Net --> init 2D, 2D, 3D (5000 iterations, lr=0.001, times 0.9 after each 100 steps)

Let e know if you had any problems.

Bardia

**Comment by payne53** [comment]
> Hi Bardia Doosti! Thank you for your excellent work. I am trying to reimplement your work and have some questions about details of pretraining on Obman dataset. The whole network can be splitted into two main parts: ResNet10+graph convolution(we name it as part1 ) and the Adaptive graph U-Net. I think there are 3 options to pretrain on Obman:
> 
> 1. Pretrain the part1 and Graph U-Net seperately. For part1, use the image from Obman as input and get 2D prediction from part1. The loss is computed between part1’s output and GT 2D points from Obman. For part2, use the GT 2D points as input, the loss is computed between Graph U-Net’s output and GT 3D points.
> 2. Pretrain the whole network end2end. That is, use image as input and the loss is computed between final 3D points output and GT 3D points.
> 3. Pretrain only the Graph U-Net.
> 
> I don’t know the exact pretraining method used in the paper, could you please clarify this? Also, could you offer other detail parameters for pretraining such as learning rate, optimizer, and epochs? Or just detailed code for pretraining can also help very much.
> 
> Thank you again for your work.

Hi, I am working on pretraining on Obman dataset. It seems Obman does not contain 8 bounding box points but only verts and faces. I wonder how you deal with these points? Do you calculate the position of bounding box according to mesh verts and faces?
Thank you very much.

**Comment by bardiadoosti** [comment]
Hi,

Good question. Here is how I handled the situation. I used Trimesh to calculate the tight bounding box of the mesh.

```
from obman.obman import ObMan
import trimesh
import numpy as np

pose_dataset = ObMan(
    '/path/to/obman',
    '/path/to/ShapeNetCore.v2',
    split='train',  # 'train', 'test', 'val
    use_cache=False,
    root_palm=False,
    mode='all',
    segment=False)

idx = 0  # index of the sample in the dataset from 0 to 141550
joints = pose_dataset.get_joints3d(idx)
vertices, faces = pose_dataset.get_obj_verts_faces(idx)
mesh = trimesh.Trimesh(vertices=vertices, faces=faces)
box = np.array(mesh.bounding_box_oriented.vertices)
```

**Comment by payne53** [comment]
> Hi,
> 
> Good question. Here is how I handled the situation. I used Trimesh to calculate the tight bounding box of the mesh.
> 
> ```
> from obman.obman import ObMan
> import trimesh
> import numpy as np
> 
> pose_dataset = ObMan(
>     '/path/to/obman',
>     '/path/to/ShapeNetCore.v2',
>     split='train',  # 'train', 'test', 'val
>     use_cache=False,
>     root_palm=False,
>     mode='all',
>     segment=False)
> 
> idx = 0  # index of the sample in the dataset from 0 to 141550
> joints = pose_dataset.get_joints3d(idx)
> vertices, faces = pose_dataset.get_obj_verts_faces(idx)
> mesh = trimesh.Trimesh(vertices=vertices, faces=faces)
> box = np.array(mesh.bounding_box_oriented.vertices)
> ```

Got it! Thanks very much!
